File: shared/types/game-state.ts
// GameState 类型定义 — 游戏运行时状态
// 对应 GAME_DESIGN.md §6.1

import type { GameScript } from './game-script';

export interface GenerationProgress {
  totalAssets: number;
  completedAssets: number;
  failedAssets: number;
  chapterStatus: Record<string, 'generating' | 'ready' | 'partial'>;
}

export interface ChoiceRecord {
  choiceNodeId: string;
  selectedOptionIndex: number;
  selectedOptionText: string;
  timestamp: number;
  chapterId: string;
  sceneId: string;
}

export interface GameState {
  saveId: string;
  gameScript: GameScript;
  currentChapterId: string;
  currentSceneId: string;
  currentNodeId: string;
  variables: Record<string, any>;
  inventory: string[];
  stats: Record<string, number>;
  choiceHistory: ChoiceRecord[];
  visitedScenes: string[];
  unlockedCGs: string[];
  generationProgress: GenerationProgress;
}

File: src-tauri/src/types/game_state.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use super::game_script::GameScript;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ChapterGenerationStatus {
    Generating,
    Ready,
    Partial,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GenerationProgress {
    pub total_assets: u32,
    pub completed_assets: u32,
    pub failed_assets: u32,
    pub chapter_status: HashMap<String, ChapterGenerationStatus>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChoiceRecord {
    pub choice_node_id: String,
    pub selected_option_index: u32,
    pub selected_option_text: String,
    pub timestamp: u64,
    pub chapter_id: String,
    pub scene_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GameState {
    pub save_id: String,
    pub game_script: GameScript,
    pub current_chapter_id: String,
    pub current_scene_id: String,
    pub current_node_id: String,
    pub variables: HashMap<String, serde_json::Value>,
    pub inventory: Vec<String>,
    pub stats: HashMap<String, f64>,
    pub choice_history: Vec<ChoiceRecord>,
    pub visited_scenes: Vec<String>,
    pub unlocked_cgs: Vec<String>,
    pub generation_progress: GenerationProgress,
}
